File: leetcode-java/src/main/java/com/wezhyn/learn/linked/RemoveNthNodeFromEndOfList.java
package com.wezhyn.learn.linked;


/**
 * @author wezhyn
 * @since 08.28.2020
 */
public class RemoveNthNodeFromEndOfList {
    public ListNode removeNthFromEnd(ListNode head, int n) {
        if (n == 0) {
            return head;
        }
        ListNode h = new ListNode(-1), slow = h, fast = h;
        h.next = head;
        for (int i = 0; i < n && fast != null; i++) {
//            前进n个
            fast = fast.next;
        }
//        n 超出范围
        if (fast == null) {
            return head;
        }
        while (fast.next != null) {
            fast = fast.next;
            slow = slow.next;
        }
        slow.next = slow.next.next;
        return h.next;
    }
}

File: leetcode-kotlin/src/main/kotlin/com/wezhyn/learn/lcof/RemoveNthNodeFromEndOfList.kt
package com.wezhyn.learn.lcof

import com.wezhyn.learn.ListNode

/**
 * 使用双指针策略，让第二个指针领先第一个指针n步
 * 给定的 N 是一直有效的
 *
 * 为了使得n =arr-1 和 n=arr-2 操作相同，添加一个伪头节点
 *
 * @author wezhyn
 * @since 06.30.2020
 *
 */
class RemoveNthNodeFromEndOfList {


    fun removeNthFromEnd(_head: ListNode?, _n: Int): ListNode? {
        val head = ListNode(-1).apply { next = _head }
        var n = _n
        var prePointer = head
        var cur = head
        while (n > 0 && head.next != null) {
            prePointer = prePointer.next!!
            n--
        }
        while (prePointer.next != null) {
            prePointer = prePointer.next!!
            cur = cur.next!!
        }
        cur.next = cur.next?.next
        return head.next
    }

}